keps/1337-preempt-within-cohort-while-borrowing/README.md

```diff
@@ -0,0 +1,484 @@
+
+|                 | A Standard    | A Best-Effort | B Standard    | B Best-Effort | Shared        |
+|-----------------|---------------|---------------|---------------|---------------|---------------|
+|nominalQuota     | 0             | 0             | 0             | 0             | 100           |
+|borrowingLimit   | 100           | 100           | 100           | 100           | 0             |
+
+The teams send their high-priority (say `>200`) jobs to their `Standard`
+ClusterQueue, and the low-priority jobs (say `<100`) to their `Best-Effort`
+ClusterQueue. Jobs sent to the `Standard` ClusterQueues should not be preempted.
+Jobs in the `Best-Effort` ClusterQueues can be preempted with respect to
+priorities.
+
+Note that in this setup every workload is borrowing, so I need to make preemption
+while borrowing enabled. Additionally, I need a mechanism to say that workloads
+from the `Standard` ClusterQueues are not preempted. A mechanism to only preempt
+jobs with priorities below a specified threshold will suffice.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Flopping between preempting pair of workloads
+
+In a setup as described in (Story 1)[#story-1] consider there are two workloads,
+one in ClusterQueue A, and one in ClusterQueue B. There is not enough resources
+in the cohort to run both workloads at the same time.
+
+If the preemption is allowed for workloads of any priorities
+(as when `.preemption.reclaimWithinCohort` equals `Any`)
+then the following flopping happens: workload A preempts workload B, then p
+preempted workload B returns to the queue and preempts workload A to make progress.
+This repeats, and none of the workloads can make any progress.
+
+In order to mitigate this risk we only enable the preemption to reclaim
+quota within cohort, for workloads requiring borrowing when such preempted
+workloads are of lower priority. This behavior is enabled by setting
+`.preemption.borrowWithinCohort.policy: LowerPriority`.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### API
+
+```golang
+// ClusterQueuePreemption contains policies to preempt Workloads from this
+// ClusterQueue or the ClusterQueue's cohort.
+type ClusterQueuePreemption struct {
+  ...
+	// borrowWithinCohort provides configuration to allow preemption within
+	// cohort while borrowing. This configuration can only be enabled when
+	// reclaimWithinCohort is not `Never`.
```

Comment:
I don't think we can make `ReclaimWithinCohort` a pointer, this is an API breaking change.

Second, if a user sets `borrowWithinCohort: LowerPriority`, then we would need to assume `reclaimWithinCohort: LowerPriority`. It makes sense, but I think we could also support setting `reclaimWithinCohort: Any`, while `borrowWithinCohort: LowerPriority`. This means allow for preemption in cohort for any priority workloads, if the preemptor is not borrowing, but only preempt lower-priority workloads if the preemptor is borrowing.

Reply:
> I don't think we can make ReclaimWithinCohort a pointer, this is an API breaking change.

IIUC, if we make `reclaimWithinCohort` a pointer, the CRD wouldn't be changed. So, we keep having backward compatibility. Maybe I discussed a similar situation with @alculquicondor. @alculquicondor If I don't understand correctly, please let me know.

Reply:
Transitioning to a pointer is somewhat backwards compatible (the yaml API doesn't change). But if there's someone that is using the golang types, they will be broken. So we should prefer not to change the types.
I think in the past we chose to change one type because it was impossible to achieve the needed semantics otherwise. In this case, the semantics could be achieved using `Never`.

However, I agree with @mimowo in that setting both policies makes sense.

Reply:
I don't think we want to use: `reclaimWithinCohort: Never` combined with `borrowWithinCohort.Policy: LowerPriority`. This would mean you can preempt in other ClusterQueues, but only if you are borrowing. I don't think this is particularly useful, I opt to reject this combination by validation.

I think the only options to make sense are:
1.  `reclaimWithinCohort: LowerPriority`, and `borrowWithinCohort.Policy: LowerPriority`; meaning you can reclaim quota from workloads in the cohort of LowerPriority, same if you are borrowing.
2.  `reclaimWithinCohort: Any`, and `borrowWithinCohort.Policy: LowerPriority`; meaning you can reclaim quota from workloads in the cohort of Any priority, however, only from lower priority workloads, if you are borrowing.

Reply:
Looking at the implementation it is easier to support only 1, then we can just reject the higher priority workloads during findCandidates. Maybe that is enough in the first iteration of the feature? WDYT @alculquicondor @tenzen-y ?

Reply:
> Maybe that is enough in the first iteration of the feature? 

I agree with you.

Thanks for clarifying this @alculquicondor @mimowo 
This makes sense.

Reply:
@mimowo Could you mention that we don't work on an opt `2` in the first iteration like this?

https://github.com/kubernetes-sigs/kueue/tree/main/keps/1284-cluster-queue-stop#notesconstraintscaveats

Reply:
I will want to support opt 2 from day 1. This is what my current implementation does. I was just hesitant for a moment that without opt 2 it is easier, but as mentioned here the solution isn't really any different https://github.com/kubernetes-sigs/kueue/pull/1405#discussion_r1429947701

Reply:
I suggest that if we decide for whatever reason during review of the implementation that it is complicated to support, then maybe we can drop the support and update KEP later.

Reply:
SGT